makiflow/experimental/newton_optimizer.py

```diff
@@ -16,33 +16,109 @@
 # along with Foobar.  If not, see <https://www.gnu.org/licenses/>.
 
 import tensorflow as tf
+# This class defines the API to add Ops to train a model.
+from tensorflow.python.ops import control_flow_ops
+from tensorflow.python.ops import math_ops
+from tensorflow.python.ops import state_ops
+from tensorflow.python.framework import ops
 
 
-class NewtonOptimizer:
-    def __init__(self, learning_rate):
-        self.learning_rate = learning_rate
-        self.grads = []
-        self.hesses = []
+def jacobian(y, x, consider_batchsize=False):
+    """
+    Computes Jacobian of the function `y` given arguments `x`.
+
+    Parameters
+    ----------
+    y : tf.Tensor
+        A Tensor of arbitrary shape.
+
+    x : tf.Tensor or tf.Variable
+        A Tensor of arbitrary shape. `x` must be the original Tensor that was used for
+        creation of the `y`. Do not perform any transformations on it before passing it.
+    """
+    y_flat = tf.reshape(y, [-1])  # [...] -> [lenght]
+    length = y_flat.get_shape().as_list()[0]
+
+    gradients = []
+    for el_ind in range(length):
+        # Computes one entry of the matrix
+        gradients += [tf.reshape(tf.gradients(y_flat[el_ind], x), shape=[-1])]
+    return tf.stack(gradients, axis=0)
+
+
+class NewtonOptimizer():
+    """
+    Implementation of Newton Optimizer.
+    See https://en.wikipedia.org/wiki/Newton%27s_method_in_optimization
+    """
+
+    NEWTON_MODE = 0
+    SGD_MODE = 0
+
+    def __init__(self, learning_rate, learning_rate_sgd=1e-4, name='NewtonOptimizer'):
+        self._name = name
+        self._lr = learning_rate
+        self._lr_sgd = learning_rate_sgd
+        self._current_state = None
+
+        # Tensor versions of the constructor arguments, created in _prepare().
+        self._lr_t = None
+        self._lr_sgd_t = None
+
+    def _prepare(self):
+        self._lr_t = ops.convert_to_tensor(self._lr, name='learning_rate')
+        self._lr_sgd_t = ops.convert_to_tensor(self._lr_sgd, name='learning_rate_sgd')
 
     def variables(self):
-        return self.grads + self.hesses
+        return [self._lr_t, self._lr_sgd_t]
 
     def _compute_update(self, var, objective):
-        grad = tf.squeeze(tf.gradients(ys=objective, xs=[var]), axis=0)
-        hess = tf.squeeze(tf.hessians(ys=objective, xs=[var]))
+        grad = tf.squeeze(tf.gradients(ys=objective, xs=var), axis=0, name=self._name + '_squeeze_grads')
+        #hess = tf.squeeze(tf.hessians(ys=objective, xs=[var]), name=self._name + '_squeeze_hess')
+
+        hess = tf.transpose(jacobian(objective, var))
```

Comment:
It might be the case that transposing the hessian is not necessary. Moreover, it can lead to an in incorrect update computation, since the gradient is represented as a row during matrix multiplication.